### ISSUE TITLE
No more IOS push notification since July 2024

### ISSUE THREAD

**Issue by GuillaumeDebernard** [binwiederhier/ntfy#1191]
# Quick description

I have NTFY running with docker-compose, behind an nginx reverse-proxy with `https` (certbot).
Since mid-July, I no longer receive notifications on IOS, which used to work fine.

The requests I send look like : `curl -d Hello https://ntfy.myDomain.com/myTopic`.

Note that reverse-proxy is used for other services and work perfectly fine for them.

I didn't see any trace log about usage limits.

# What's still working

- Notifications on Android phones
- Notifications on WebApp (Chrome & Safari at least)
- Notifications arrives on IOS if you refresh the app (by pulling down)
- Push notification to IOS with `https://ntfy.sh/myTopic`

# What I tried, and didn't work

- Run a previous ntfy version
- Prune docker image and rebuild it
- Switch from `server.yml` to env variables for configuration
- Switch from `https://myDomain.com/services/path/to/MyTopic` to subdomain : `https://ntfy.myDomain.com/myTopic` (It allow me to acces webapp by the way)
- Uninstall/reinstall the app + remove app from Icloud backup
- Restart Iphone
- Unsuscribe/resuscribe to topics
- A lot of nginx configuration for the redirections

# My configurations

- `NTFY_BASE_URL` match exactly the default server I set up in my app.

## docker-compose.yaml file

``` dockerfile
ntfy:
  image: binwiederhier/ntfy:latest
  container_name: ntfy_container
  command:
    - serve
  environment:
    TZ: "Europe/Paris"
    NTFY_BASE_URL: https://${SUB_NTFY}
    NTFY_LISTEN_HTTP: :${NTFY_PORT}
    NTFY_CACHE_FILE: /var/cache/ntfy/cache.db
    NTFY_ATTACHMENT_CACHE_DIR: /var/cache/ntfy/attachments
    NTFY_UPSTREAM_BASE_URL: https://ntfy.sh
    NTFY_LOG_LEVEL: warning
    NTFY_LOG_FORMAT: text
    NTFY_LOG_FILE: var/ntfy/logs.log
    NTFY_KEEPALIVE_INTERVAL: 45s
    NTFY_BEHIND_PROXY: true
    NTFY_WEB_ROOT: /app
  volumes:
    - ./utils/ntfy/cache:/var/cache/ntfy
    - ./utils/ntfy:/var/ntfy
  ports:
    - "${NTFY_PORT}:${NTFY_PORT}"
   # Work the same if using port 80
  networks:
    - main
```

## Nginx default.conf.template file

``` nginx
server {
  listen ${HTTPS_DEFAULT} ssl http2;
  listen [::]:${HTTPS_DEFAULT} ssl http2;

  server_name ${SUB_NTFY};

  ssl_certificate /etc/letsencrypt/live/${DOMAIN}/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/${DOMAIN}/privkey.pem;

  set $upstream_ntfy "ntfy:${NTFY_PORT}";

  resolver 127.0.0.11 valid=1m;

  location / {

    rewrite /(.*) /$1 break;
    proxy_pass http://$upstream_ntfy/$1$is_args$args;
    proxy_http_version 1.1;

    proxy_set_header Host $http_host;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

    proxy_connect_timeout 3m;
    proxy_send_timeout 3m;
    proxy_read_timeout 3m;

    client_max_body_size 0;
  }
}
```

What i really don't understand is why it stopped working. I should have read a hundred times all the configuration page from `ntfy.sh` without finding any problems in my configuration.  
Since I didn't see any new post about that issue I assumed it was an nginx configuration issue, but every other services I host, and ntfy usage (webapp, android) are still working.
I upgrade recently IOS 17 to 18 without any improvement.

I'm running out of ideas, if anyone is facing the same issue, join the club ;)

I'd also like to take this opportunity to congratulate you on your great work!

**Comment by a1ad** [comment]
Yeah, same here

**Comment by wunter8** [comment]
In general, notifications to iOS are tricky. I haven't heard anything specific about updates to iOS 18 or Nginx causing problems, though. It's definitely strange that things were working and then stopped (at least for OP).

You'll have better luck with notifications on iOS through the PWA. You'll need to set up web push keys on your self-hosted server (and have the server behind a domain with a TLS cert that is publicly trusted (e.g., not self-signed)), but it's easy to set up. With the PWA, notifications should arrive reliably, and you'll have more features than the native iOS app (like markdown formatting and embedded image previews).

Here's the list of things I tell people to do to get iOS notifications working, so you can double-check that everything is in order (but again, if it was working and stopped without any sort of config change, I don't know what would cause that):
1. open a browser to the web app of your ntfy instance and copy the URL (including "http://" or "https://", your domain or IP address, and any ports, and excluding any trailing slashes)
2. put the URL you copied in the ntfy `base-url` config in server.yml or NTFY_BASE_URL in env variables
3. put the URL you copied in the default server URL setting in the iOS ntfy app
4. set `upstream-base-url` in server.yml or NTFY_UPSTREAM_BASE_URL in env variables to "https://ntfy.sh"

**Comment by plittlefield** [comment]
Wow, very detailed reply many thanks.

I will try that tomorrow and report back.

**Comment by plittlefield** [comment]
Hi @wunter8 I have news.

The addition of the variable NTFY_UPSTREAM_BASE_URL to my docker-compose.yml now makes a Notification appear in the Notification Centre of my iPhone.

However, this is still not quite good enough for me, because I need it to appear on the Lock Screen ... just like Pushover does.

Also, what is PWA ?

**Comment by plittlefield** [comment]
OK, I take that back!

I have just received a ntfy notification on my lock screen and my watch!

Happy days!

**Comment by a1ad** [comment]
@plittlefield with the app from the store? Mine is still not working

@wunter8 can we use "server.yml" and the docker ENVs at the same time?

**Comment by plittlefield** [comment]
OK, I might try that tomorrow.

What do I have to install on the phone for this work or does it just use the existing iOS app?

**Comment by wunter8** [comment]
After setting up the web push keys, load the self-hosted web app in Safari and then go to Share > "add to homescreen"

**Comment by plittlefield** [comment]
Oooo, will that give me nice count badges?!

**Comment by wunter8** [comment]
Background notifications for PWAs on iOS have only been supported since iOS 16.4 (March 2023).

I'm not sure about the count badges 🤷‍♂️ (I don't use iOS)

**Comment by wunter8** [comment]
Oh, I know what they are. I just don't know if PWAs support it in general. And I don't know if the ntfy PWA supports them specifically.

(Or were you showing us the ntfy app doesn't have a badge meaning they're apparently not supported? I also can't tell if that's the ntfy app or the ntfy PWA)

**Comment by plittlefield** [comment]
A bit of both - showing you the unread badge count on WhatsApp and that ntfy does not support that feature yet.

**Comment by plittlefield** [comment]
Web push configured and seems to be OK but after putting the Safari bookmark on the iOS home screen (and then a login to my ntfy account) I no only do not get new notifications but also no unread badges.

Oh well.

I shall stick to the actual app for now.

**Comment by plittlefield** [comment]
I’m trying Chrome on iOS as well - it shows the notice about adding notifications - 
![IMG_1493](https://github.com/user-attachments/assets/b5b67f19-bb60-4520-be48-3ebf454e0510)

**Comment by wunter8** [comment]
When you click on the ntfy-web icon on your homescreen, does it just load the web app in Safari or does it launch a PWA in a standalone screen (without the Safari URL bar on top and stuff)?

**Comment by plittlefield** [comment]
I think it’s a PWA … what do you think?

![IMG_1496](https://github.com/user-attachments/assets/0711aa29-1d8a-4bb7-81e1-300a7ac95a0e)

![IMG_1497](https://github.com/user-attachments/assets/5d06b042-be49-4c37-b21b-eff80e7867a6)

… getting there, but of course it’s the actual iOS app that’s showing me notifications on my “Notification Centre” and not the web app.

At least I think so!

Did you want me to uninstall the actual App to see ?

**Comment by plittlefield** [comment]
Okay, that's uninstalled ... drum roll for later today :-)

**Comment by plittlefield** [comment]
Well blow me down with a feather!


![IMG_1506](https://github.com/user-attachments/assets/a8760bd2-e7d1-4b22-92fb-ee497e4918ea)

**Comment by svewag** [comment]
I can confirm that the notifications on iOS are working just fine in the PWA, but not in the app.

**Comment by chlorine3545** [comment]
@wunter8 Sorry to interrupt. What is NTFY_UPSTREAM_BASE_URL? Do notifications from our self-hosted server need to be forwarded to the official ntfy server?

Also, if I may ask politely, is it because the official ntfy server has APNS configured, whereas our self-hosted server does not?

I'm a novice in this area, so apologies for any mistakes.

**Comment by wunter8** [comment]
@chlorine3545 You are correct. Here's some info: https://docs.ntfy.sh/config/#ios-instant-notifications

Let me know if you have more questions

**Comment by chlorine3545** [comment]
First of all, thanks for your reply @wunter8.

Not only Ntfy, but all these unified push can't work properly on iOS, we have to use APNs, which is a shame to some extent. Maybe I should get an Android phone and flash it with LineageOS or something.

**Comment by lbeier** [comment]
Hey folks! First of all, thanks for this amazing project! 🥇 

I'm also having issues with iOS push notifications. I've set both `base-url` and `upstream-base-url` and I believe they are correct.

Not sure what else I'm missing though. It's likely be something really stupid, but I can't really understand what it is.

Here is my `docker-compose.yml`:
```yaml
version: "2.1"
services:
  ntfy:
    image: binwiederhier/ntfy:latest
    container_name: ntfy
    command:
      - serve
    environment:
      - TZ=Europe/Madrid
    volumes:
      - lib:/var/lib/ntfy
      - config:/etc/ntfy
    ports:
      - 31209:80 # I have something else running on port 80, so I generated a random port for host
    restart: unless-stopped
    labels:
      - "com.centurylinklabs.watchtower.enable=true"

volumes:
  lib: {}
  config: {}
```

And here is my `server.yml`:
```yaml
base-url: "https://ntfy.<MY-HOMELAB-HOST>"
upstream-base-url: "https://ntfy.sh"

auth-file: "/var/lib/ntfy/auth.db"
auth-default-access: "deny-all"
enable-login: true

behind-proxy: true

cache-file: "/var/lib/ntfy/cache.db"

attachment-cache-dir: "/var/lib/ntfy/attachments"
attachment-total-size-limit: "5G"
attachment-file-size-limit: "15M"
attachment-expiry-duration: "3h"
visitor-attachment-total-size-limit: "100M"
visitor-attachment-daily-bandwidth-limit: "500M"
```

Some additional info that might be relevant:
- If I open my NTFY server page in a browser in my laptop, I get instant notifications
- My NTFY instance is behind a Nginx Proxy Manager, so the `behind-proxy` needs to be set to true as per my understanding.
- On iOS NTFY app:
    - The default server is set to my `https://ntfy.<MY-HOMELAB-HOST>"` with `https://` and no trailing `/`.
    - I've logged in with the admin user I created.
    - I can see the messages if I manually refresh the notifications page, the issue is only with push notifications
    - notifications are set to be delivered instantly
    - is allowed to run in background
- I have network wide adblocking (Adguard Home), but I don't see anything being blocked by Adguard that is related to NTFY container.

**Comment by lbeier** [comment]
So, I've installed the NTFY app on my iPad and the notifications it's working on the iPad.

Both are in public beta.

iOS 18.1 (22B83)
iPadOS 18.1 (22B82)

There was a pending update on my iPad. Just installed it (iPadOS 18.1 (22B83)) and the notifications are still working.

The issue seems to be isolated to iOS only. 🤔

**Comment by wunter8** [comment]
Hmm. I haven't heard of someone having issues on iOS but not on iPadOS. That seems to suggest your ntfy server is configured correctly. I've heard some people say reinstalling the ntfy app can help. But I'm not sure what else to try.

I generally recommend people use the PWA on iOS, though. It has more features than the native app and more reliable notifications. You need to configure web push to get it to work: https://docs.ntfy.sh/config#web-push

**Comment by lbeier** [comment]
Hey @wunter8!

Thanks for the suggestion! I've setup and I'm getting notifications on the pwa now.

Will still investigate a bit mor about the iOS specific issue and get back here if I find something. But for now, the pwa is working just fine.

**Comment by plittlefield** [comment]
I'd like to chime in and say that while the PWA is working great, there does not seem to be a synchronisation between the PWA app running on my phone and the web site in a browser tab.

What I mean by this is that when I get a notification it appears on both devices, BUT when I read or read and delete from either of the devices, the other device does not appear the same ... indeed all notifications remain UNREAD.

**Comment by wunter8** [comment]
@plittlefield That is the expected behavior right now. The state of individual messages is not synchronized between devices/clients

**Comment by soffio** [comment]
I set up a self-hosted a few days ago and it worked fine. But yesterday I couldn't receive notifications anymore. Today I reinstalled the app and it worked again. It's so weird.

**Comment by Vhaelan** [comment]
> Hi @wunter8 I have news.
> 
> The addition of the variable NTFY_UPSTREAM_BASE_URL to my docker-compose.yml now makes a Notification appear in the Notification Centre of my iPhone.
> 
> However, this is still not quite good enough for me, because I need it to appear on the Lock Screen ... just like Pushover does.
> 
> Also, what is PWA ?

I just wanted to add that setting the NTFY_UPSTREAM_BASE_URL environment variable also fixed the issue for me, even though it was also set in server.yml before. The symptoms were the same for me, notifications just stopped coming during the summer, without any changes to my config, but it seems the env variable solves the issue.

**Comment by aryapramudika** [comment]
I have same problem, after change the _**upstream-base-url to**_  https://ntfy.sh now notification is worked.

here my server.yml

```
auth-default-access: "deny-all"
auth-file: "/var/lib/ntfy/user.db"
base-url: "https://ntfy.yourdomain.net"
listen-http: ":2586"
upstream-base-url: "https://ntfy.sh"
visitor-request-limit-burst: 100
```

**Comment by Victor-Buendia** [comment]
> have same problem, after change the _**upstream-base-url to**_ https://ntfy.sh now notification is worked.

Same here! This was the only change needed and the iOS App notifications are now back.

**Comment by kilianc** [comment]
downloading the official app, adding a test topic and tapping sending test does not trigger a notification but refreshing the page shows it was published. I tried pretty much everything in the settings but nothing was helpful.

I understand this is a free product, but I wanted to mention that at this point in time the iOS notification flow, as shown in the homepage does not work with the free tier on the latest iOS (18.2).

**Comment by Snake16547** [comment]
Also wanna jump in here.
Configured everything like you guys. Push notifications coming through the only thing I'm missing is the little red number over the native ntfy app icon. Is this even possible?

**Comment by jdoncouse** [comment]
> > The addition of the variable NTFY_UPSTREAM_BASE_URL to my docker-compose.yml now makes a Notification appear in the 

Chiming in as adding this to my env variables fixed the notifications for my iPhone

**Comment by Amwagner45** [comment]
None of the above worked for me I added the base-upstream-url in the yaml file and set as an environment variable. When I go to the app and scroll down I can see new notifications but it isn't creating a banner or add to my Notification Center on iOS. Seems like the only other option is the PWA but that will need to be another weekend project to change.

**Comment by wunter8** [comment]
@Amwagner45 These are the things you need to do to get iOS push notifications to work:
1. open a browser to the web app of your ntfy instance and copy the URL (including "http://" or "https://", your domain or IP address, and any ports, and excluding any trailing slashes)
2. put the URL you copied in the ntfy `base-url` config in server.yml or NTFY_BASE_URL in env variables
3. put the URL you copied in the default server URL setting in the iOS ntfy app
4. set `upstream-base-url` in server.yml or NTFY_UPSTREAM_BASE_URL in env variables to "https://ntfy.sh" (without a trailing slash)

**Comment by amcmanu3** [comment]
I'm also experiencing this issue despite setting NTFY_UPSTREAM_BASE_URL

![Image](https://github.com/user-attachments/assets/dc6c9704-99ff-4e7e-8c45-2f2b641a0da9)

**Comment by adamcravas** [comment]
@wunter8 -- Is the PWA iOS Push notifications supported if we're not self hosting and just using the public https://ntfy.sh topics?  I've run into similar issues but am not hosting my own NTFY.

**Comment by wunter8** [comment]
Yes, the public instance has web push configured, so you can use it as a PWA. Just go to the web app (https://ntfy.sh/app) and "Save to homescreen"

**Comment by adamcravas** [comment]
@wunter8  -- Thanks for the quick reply!  I've done that and I can see the red bubble number increase as I push messages from my desktop.  However, I have to have the web app open for this to happen.  I don't receive any push notifications.  Do you happen to have any experience in trouble shooting for that?

**Comment by adamcravas** [comment]
Oh I've got it now! For posterity -- the key (for me) was to Save to home screen from https://ntfy.sh/app (as wunter8 mentioned exactly above).  I originally had https://ntfy.sh.

Thanks for all the help @wunter8.

**Comment by garymoon** [comment]
I tried all the fixes I saw in this thread and no glory. Then I noticed my `base-url` was `http` instead of `https` 🤦‍♂️😂

Thanks to everyone here sharing fixes, and to binwiederhier for ntfy 💙

**Comment by binwiederhier** [comment]
We've improved the iOS app in the last few releases, and we're hoping that some of these issues are resolved. Please open a new issue and describe things in detail if things are still not working. To receive help/support, please use the Discord/Matrix chat instead.
